### ISSUE TITLE
What format does Carthage expect for binary release zip files?

### ISSUE THREAD

**Issue by dpassage** [Carthage/Carthage#1056]
Background: I'm trying to adapt our build system to directly build a binary release for attachment to a private Github repo, such that Carthage can manage the dependency. Currently, the zip file's contents look like this:

```
Carthage/
   Build/
      iOS/
         Foo.framework
```

I can add a reference to the `Cartfile`; it downloads the binary release without complaint, but the framework doesn't appear under the `Carthage` directory of my project.

I've tried a few combinations, including making sure the name of the framework matches the name of the repo (`Foo` in the example above) with no luck. I've compared our .zip file with the file from an open source project that does work for binary releases; things look pretty much the same.

Any guidance available?

**Comment by mdiep** [comment]
I don't remember exactly what Carthage looks for, but I think @ikesyo will remember the details.

**Comment by ikesyo** [comment]
Carthage would do the following:

1. Picks up any frameworks which does not have nested frameworks in a archive: https://github.com/Carthage/Carthage/blob/0.11/Source/CarthageKit/Project.swift#L718-L728
2. Copy a framework to the appropriate platform folder (`Carthage/Build/$platform`). The platform will be determined by the `DTSDKName` value of the `Info.plist` file in the framework (`Resources/Info.plist` for Mac frameworks, `Info.plist` for other platforms): https://github.com/Carthage/Carthage/blob/0.11/Source/CarthageKit/Project.swift#L437-L449

In conclusion, there is no expected directory layout Carthage will do. The problem would be that your frameworks

- have nested frameworks in `Frameworks` folder.
- don't contain `Info.plist` file.
- the plist file does not have `DTSDKName` value.

**Comment by dpassage** [comment]
Thanks, @ikesyo! I'll let you know how it goes.

**Comment by ikesyo** [comment]
@dpassage Do you have any updates on this?

**Comment by dpassage** [comment]
@ikesyo Yes, and thanks again for the info.

Looks like Carthage can't find the plist file:

```shell
$ ls -l Info.plist Resources/Info.plist
ls: Info.plist: No such file or directory
ls: Resources/Info.plist: No such file or directory
```

It looks like the framework in question is putting its `Info.plist` file down in a bundle directory:

```shell
$ find . -name Info.plist
./Versions/A/Resources/FooBundle.bundle/Info.plist
```

I don't have enough framework experience to know how weird this is, but Xcode seems to consume it perfectly happily. Unfortunately, the framework in question is closed source, so I can't share it directly. I can try to see if I can share how it's being built, or an example of the build.

**Comment by mdiep** [comment]
I wonder if we should be using `NSBundle` and `NSBundle.infoDictionary` to get that information, rather than trying to guess the location of `Info.plist`.

**Comment by ikesyo** [comment]
@dpassage It seems that the plist file is not for `Foo.framework`, but for `FooBundle.bundle`. So the framework does not have any its `Info.plist`.

I tried `NSBundle.infoDictionary` for a framework with that structure and that returned `nil`.

**Comment by dpassage** [comment]
Well, that'd do it. I'll report this upstream. Thanks for your help!
